File: GAMEOFLIFE/World.cpp
#include "World.h"
#include "Organism.h"
#include "Human.h"
#include "Grass.h"
#include "Dandelion.h"
#include "Guarana.h"
#include "DeadlyNightshade.h"
#include "SosnowskyHogweed.h"
#include "Wolf.h"
#include "Sheep.h"
#include "Fox.h"
#include "Turtle.h"
#include "Antelope.h"
#include <iostream>
#include <string>
#include <fstream>

#define NOTHING 13
#define OCCUPIED 1

World::World(int right, int bot){
    this->right = right;
    this->bot = bot;
    round = 0;
    logs = 0;
    animals = 0;
    human = nullptr;
}

World* World::current = nullptr;

World* World::GetCurrentWorld() {
    return current;
}

World* World::GetCurrent(int right, int bot) {
    if (current != nullptr)
        delete current;
    current = new World(right, bot);
        return current;
}

void World::DrawWorld() const {

    mvprintw(0, 3, "GAME OF LIFE, 2024, Tamara Mruk, s197584");
    mvprintw(3, 20, "OO");
    move(3, 22);
    hline('-', right);
    mvprintw(3, right+left, "OO");
    move(4, 20);
    vline('|', bot);
    mvprintw(4+bot, 20, "OO");
    move(4, 21);
    vline('|', bot);
    move(4+bot, 22);
    hline('-', right);
    mvprintw(4+bot, 22+right, "OO");
    move(4, 22+right);
    vline('|', bot);
    move(4, 23+right);
    vline('|', bot);
    mvprintw(4, 55, "Press [R] to move to the next round");
    mvprintw(6, 55, "Press [X] to use burning flames as the human");
    mvprintw(8, 55, "Use WASD to prepare a move for the human");
    mvprintw(10, 55, "Press [O] to save game, and [L] to load game");
    mvprintw(14, 55, "LAST EVENT:");

    mvprintw(3, 2, "ROUND: %d", round);
    mvprintw(5, 2, "ORGANISMS: %d", animals);

    for (Organism* o : organisms)
    {
        if(o->GetStrength()>-1)
		    o->Draw();
    }

    refresh();
}

void World::DoRound() {

    round++;

    for(int i=4; i<4+bot; i++){
        for(int j=22; j<22+right; j++){
            attron(COLOR_PAIR(NOTHING));
            mvaddch(i, j, ' ');
            attroff(COLOR_PAIR(NOTHING));
        }
    }
    attron(COLOR_PAIR(NOTHING));
    mvaddch(3, 9, ' ');
    attroff(COLOR_PAIR(NOTHING));
    attron(COLOR_PAIR(NOTHING));
    mvaddch(3, 10, ' ');
    attroff(COLOR_PAIR(NOTHING));
    attron(COLOR_PAIR(NOTHING));
    mvaddch(5, 13, ' ');
    attroff(COLOR_PAIR(NOTHING));
    attron(COLOR_PAIR(NOTHING));
    mvaddch(5, 14, ' ');
    attroff(COLOR_PAIR(NOTHING));
    attron(COLOR_PAIR(NOTHING));
    mvaddch(7, 18, ' ');
    attroff(COLOR_PAIR(NOTHING));
    attron(COLOR_PAIR(NOTHING));
    mvaddch(7, 19, ' ');
    attroff(COLOR_PAIR(NOTHING));

    std::sort(organisms.begin(), organisms.end(), Organism::CompareInitiativeAndAge);

    for (Organism* o : organisms) {
        if (o->GetStrength() > -1) {
            o->Action();
        }
        else {
            break;
        }
	}
}

void World::AddOrganism(Organism* organism) {
    organisms.push_back(organism);
    SetAnimals(GetAnimals()+1);
}

void World::SpawnOrganisms() {

    std::random_device rd;
    std::mt19937 eng(rd());
    std::uniform_int_distribution<> distributionHorizontal(left, left+right-1);
    std::uniform_int_distribution<> distributionVertical(top, top+bot-1);

    int randH=distributionHorizontal(eng), randV=distributionVertical(eng);

    int spaces[20][20]={0};

    spaces[top+(bot/2)][left+(right/2)]=OCCUPIED;


    human = new Human(left+(right/2), top+(bot/2));
    

    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Grass(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Dandelion(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Guarana(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new DeadlyNightshade(randH, randV);
    

    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new SosnowskyHogweed(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Wolf(randH, randV);

    // while (spaces[randV-top][randH-left]==OCCUPIED) {
    //     randH = distributionHorizontal(eng);
    //     randV = distributionVertical(eng);
    // }
    // spaces[randV-top][randH-left]=OCCUPIED;
    // new Wolf(randH, randV);



    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Sheep(randH, randV);

    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Sheep(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Fox(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Turtle(randH, randV);


    while (spaces[randV-top][randH-left]==OCCUPIED) {
        randH = distributionHorizontal(eng);
        randV = distributionVertical(eng);
    }
    spaces[randV-top][randH-left]=OCCUPIED;
    new Antelope(randH, randV);
}

int World::GetTop() {
    return top;
}

int World::GetLeft() {
    return left;
}

int World::GetBot() {
    return bot;
}

int World::GetRight() {
    return right;
}

int World::GetRound() {
    return round;
}

int World::GetLogs() {
    return logs;
}

void World::SetLogs(int amount){
    logs = amount;
}

void World::SetAnimals(int amount){
    animals = amount;
}

int World::GetAnimals(){
    return animals;
}

Organism* World::GetOrganism(pair<int,int> pos)
{
    if(pos.first < GetLeft() || pos.first >= (GetRight()+GetLeft()-1) || pos.second < GetTop() || pos.second >= (GetBot()+GetTop()-1)) {
        return nullptr;
    }
    for (Organism* o : organisms) {
        if (o->GetX() == pos.first && o->GetY() == pos.second) {
			return o;
		}
	}
	return nullptr;
}

Organism* World::GetHuman() {
    return human;
}

std::vector<Organism*> World::GetOrganisms() {
    return organisms;
}

void World::ClearLogs(){
    for(int i=16; i<40; i++){
        for(int j=55; j<130; j++){
            attron(COLOR_PAIR(NOTHING));
            mvaddch(i, j, ' ');
            attroff(COLOR_PAIR(NOTHING));
        }
    }
}

void World::AddLog(Organism* source, string log, int cut){ 
    if (GetLogs()>9){
        ClearLogs();
        SetLogs(0);
    }
    string creature = typeid(*source).name();
    log = creature + " " + log;
    if(cut==2){
        log = log.substr(2,log.length());
    }
    else{
        log = log.substr(1,log.length());
    }
    mvaddstr(16+(GetLogs()*2), 55, log.c_str());
    SetLogs(GetLogs()+1);
}

void World::AddInfo(string log){
    if (GetLogs()>9){
        ClearLogs();
        SetLogs(0);
    }

    mvaddstr(16+(GetLogs()*2), 55, log.c_str());
    SetLogs(GetLogs()+1);
}

void World::Save(){

    ofstream file;

    file.open("save.txt");

    file << right <<endl;
    file << bot <<endl;
    file << round <<endl;
    
    for (Organism* o : organisms){
        if (o->GetStrength() > -1) {
			file << o->Data() << endl;
		}
	}
    file.close();

    AddInfo("Saved the game");
}

World* World::Load(){

    delete current;

    current = nullptr;

    ifstream file;
    file = ifstream("save.txt");

    int right, bot, round, ability, age;

    string line;

    getline(file, line);
    right = stoi(line);

    getline(file, line);
    bot = stoi(line);

    getline(file, line);
    round = stoi(line);

    current = new World(right, bot);
    current->round = round;
    current->logs = 0;
    current->animals = 0;

    while (getline(file, line, '|')) {
        Organism* o = nullptr;
        if (line.size()<2){
            continue;
        }

        string name = line.substr(line.find(' ')+1, line.length());
        getline(file, line, '|');
        int x = stoi(line);
        getline(file, line, '|');
        int y = stoi(line);
        getline(file, line, '|');
        int strength = stoi(line);
        if (name == "5Human") {
            o = new Human(x, y);
            current->human = o;
            getline(file, line, '|');
            age = stoi(line);
            getline(file, line);
            int ability = stoi(line);
            ((Human*)o)->SetAbility(ability);
        }
        else if(name=="4Wolf"){
            getline(file, line);
            age = stoi(line);
            o = new Wolf(x,y);
        }
        else if(name=="5Sheep"){
            getline(file, line);
            age = stoi(line);
            o = new Sheep(x,y);
        }
        else if(name=="3Fox"){
            getline(file, line);
            age = stoi(line);
            o = new Fox(x,y);
        }
        else if(name=="6Turtle"){
            getline(file, line);
            age = stoi(line);
            o = new Turtle(x,y);
        }
        else if(name=="8Antelope"){
            getline(file, line);
            age = stoi(line);
            o = new Antelope(x,y);
        }
        else if(name=="5Grass"){
            getline(file, line);
            age = stoi(line);
            o = new Grass(x,y);
        }
        else if(name=="9Dandelion"){
            getline(file, line);
            age = stoi(line);
            o = new Dandelion(x,y);
        }
        else if(name=="7Guarana"){
            getline(file, line);
            age = stoi(line);
            o = new Guarana(x,y);
        }
		else if(name=="16DeadlyNightshade"){
            getline(file, line);
            age = stoi(line);
            o = new DeadlyNightshade(x,y);
        }
		else if(name=="16SosnowskyHogweed"){
            getline(file, line);
            age = stoi(line);
            o = new SosnowskyHogweed(x,y);
        }
        o->SetStrength(strength);
        o->SetAge(age);
        current->ClearLogs();
	}
    file.close();
    return current;
}

World::~World(){
    for (auto o : organisms) {
        delete o;
    }
}

File: GAMEOFLIFEJAVA/gameoflife/World.java
package gameoflife;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.stream.Collectors;

import gameoflife.organisms.Organism;
import gameoflife.organisms.animals.Human;
import gameoflife.structs.Vector;
import gameoflife.structs.RandomGenerator;

public abstract class World {

    int height;
    int width;
    int round;
    Human human;
    String logs;
    ArrayList<Organism> organisms;
    Vector initialPosition;
    static World currentInstance = null;

    static public World GetInstance(int width, int height) {
        currentInstance = new RecWorld(width, height);
        return currentInstance;
    }

    static public World GetInstance() {
        return currentInstance;
    }

    public void DoRound(){
        logs = "";
        ++round;
        organisms = new ArrayList<>(organisms.stream().filter(o -> o.GetStrength() >= 0).collect(Collectors.toList()));
        organisms.sort(Organism::Compare);
        int size = organisms.size();
        for (int i = 0; i < size; i++) {
            if(organisms.get(i).GetStrength()>=0)
                organisms.get(i).Action();
        }
    }

    public void Log(String message){
        logs+=message+"\n";
    }

    public void SetRound(int amount) {
        round = amount;
    }

    public Boolean OnBoard(Vector position){
        return position.GetX()>=0 && position.GetX()<width && position.GetY()>=0 && position.GetY()<height;
    }
    public void AddOrganism(String name, Vector position){
        Organism o = Organism.Spawn(this, name, position);
        organisms.add(o);
        o.previousPosition = o.position;
        Log(o.getClass().getSimpleName() + " appeared at " + position.ToString());
    }

    public Organism GetOrganism(Vector position) {
        return organisms.stream().filter(o -> o.GetPos().IsEqual(position)).findFirst().orElse(null);
    }

    public abstract Vector[] GetPossibleMoves();

    void SpawnOrganisms(){
        RandomGenerator r = new RandomGenerator();
        AddOrganism("animals.Human", new Vector(r.random(0,width-1), r.random(0,height-1)));
        human = (Human) organisms.get(0);
        AddOrganism("animals.Sheep", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("animals.Sheep", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("animals.Wolf", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("animals.Fox", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("animals.Turtle", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("animals.Antelope", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("plants.Grass", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("plants.Dandelion", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("plants.DeadlyNightshade", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("plants.Guarana", new Vector(r.random(0,width-1), r.random(0,height-1)));
        AddOrganism("plants.SosnowskyHogweed", new Vector(r.random(0,width-1), r.random(0,height-1)));
    }

    public Human GetHuman(){
        return human;
    }

    public void SaveGame(){

        String filePath = "save.txt";

        try {
            FileWriter writer = new FileWriter(filePath);
            
            writer.write(height + "|" + width + "|" + round + "\n");

            for (Organism organism : organisms) {
                String name = organism.getClass().getCanonicalName();
                if(name=="gameoflife.organisms.animals.Human"){
                    writer.write(name + "|" + organism.position.GetX() + "|" + organism.position.GetY() + "|" + organism.GetStrength() + "|" + organism.GetAge() + "|" + human.ability + "\n");
                }
                else {
                    writer.write(name + "|" + organism.position.GetX() + "|" + organism.position.GetY() + "|" + organism.GetStrength() + "|" + organism.GetAge() + "\n");
                }
            }
            Log("Game saved successfully!");
            writer.close();
        } catch (IOException e) {
            Log("An error occurred while saving the game data: " + e.getMessage());
            e.printStackTrace();
        }
    }

    public void LoadGame() {

        String[] data;
        String filePath = "save.txt";
        int x, y, strength, age, ability;
        Vector vector;
        String line;

        try {
            BufferedReader reader = new BufferedReader(new FileReader(filePath));

            line = reader.readLine();
           
            while ((line = reader.readLine()) != null) {

                data = line.split("\\|");

                String fullname = data[0];
                String[] parts = fullname.split("\\.");
                String name;

                name = parts[parts.length - 2] + "." + parts[parts.length - 1];

                x = Integer.parseInt(data[1]);
                y = Integer.parseInt(data[2]);
                vector = new Vector(x, y);
                strength = Integer.parseInt(data[3]);
                age = Integer.parseInt(data[4]);

                AddOrganism(name, vector);
                Organism organism = GetOrganism(vector);

                if (organism instanceof Human) {

                    human = (Human) organism;
                    human.SetStrength(strength);
                    human.SetAge(age);
                    ability = Integer.parseInt(data[5]);
                    human.SetAbility(ability);
                }
            }
            reader.close();
        } catch (IOException e) {
            System.out.println("An error occurred while loading the game data: " + e.getMessage());
            e.printStackTrace();
        }
        Log("Game loaded successfully!");
    }
}
